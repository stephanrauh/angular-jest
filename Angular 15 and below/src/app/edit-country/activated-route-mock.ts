import { ActivatedRoute } from '@angular/router';
import { BehaviorSubject } from 'rxjs';

export const mockActivateRouteValue = {
  data: new BehaviorSubject<any>({data: 'some additional data'}),
  params: new BehaviorSubject<any>({country: 'Spain'}),
  snapshot: {
    url: [
      {
        path: 'country',
      },
      {
        path: 'Spain',
      },
    ],
  }
} as unknown as ActivatedRoute; // fancy typecast allows us to omit unused values

export const mockActivatedRoute = {
  provide: ActivatedRoute,
  useValue: mockActivateRouteValue
};
