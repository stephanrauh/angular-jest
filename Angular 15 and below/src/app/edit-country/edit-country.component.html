<p-confirmDialog
  key="confirm-save-country"
  header="Save country"
  message="Do you want to save the country?"
  acceptLabel="I might!"
  rejectLabel="No. That's too much work."
></p-confirmDialog>
<div class="padding">
  <form [formGroup]="formGroup">
    <div class="p-grid">
      <h2 class="p-col-12">{{ (country$ | async)?.name }}</h2>

      <div class="p-col-2">Name</div>
      <div class="p-col-10">
        <input name="name" formControlName="name" pInputText />
      </div>
      <small
        class="error p-col-12 p-offset-2"
        *ngIf="formGroup.get('name')?.hasError('required')"
      >
        The field is required.
      </small>

      <div class="p-col-2">Population</div>
      <div class="p-col-10">
        <input name="population" formControlName="population" pInputText />
      </div>
      <small
        class="error p-col-12 p-offset-2"
        *ngIf="formGroup.get('population')?.hasError('required')"
      >
        The field is required.
      </small>

      <div class="p-col-2">Region</div>
      <div class="p-col-10">
        <input name="region" formControlName="region" pInputText />
      </div>
      <small
        class="error p-col-12 p-offset-2"
        *ngIf="formGroup.get('region')?.hasError('required')"
      >
        The field is required.
      </small>
      <small
        class="error p-col-12 p-offset-2"
        *ngIf="formGroup.get('region')?.hasError('minlength')"
      >
        The minimum length for this field is
        {{
          formGroup.get("region")?.errors?.minlength.requiredLength
        }}
        characters.
      </small>

      <div class="p-col-2">Flag</div>
      <div class="p-col-10">
        <input name="flag" formControlName="flag" pInputText />
      </div>
      <small
        class="error p-col-12 p-offset-2"
        *ngIf="formGroup.get('flag')?.hasError('required')"
      >
        The field is required.
      </small>
    </div>
    <button (click)="save()" [disabled]="!(canSave$ | async)" pButton>
      Save
    </button>
  </form>
</div>
