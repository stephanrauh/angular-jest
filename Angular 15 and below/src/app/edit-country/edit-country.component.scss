
.padding {
  padding-left: 8px;
  padding-right: 8px;
  padding-bottom: 20px;
  padding-top: 20px;
}

input {
  min-width: 30rem;
}

.error {
  color: red;
  margin-top: -13px;
}
