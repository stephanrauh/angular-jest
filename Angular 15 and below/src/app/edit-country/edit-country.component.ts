import { Component, OnInit } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { ActivatedRoute } from '@angular/router';
import { ConfirmationService } from 'primeng/api';
import { Observable } from 'rxjs';
import { map, switchMap, tap } from 'rxjs/operators';
import { ContinentService } from '../continent-service/continent.service';
import { Country } from '../continent-service/country';

@Component({
  selector: 'app-edit-country',
  templateUrl: './edit-country.component.html',
  styleUrls: ['./edit-country.component.scss'],
})
export class EditCountryComponent implements OnInit {
  public country$!: Observable<Country | undefined>;

  public formGroup: FormGroup = new FormGroup({
    name: new FormControl('', Validators.required),
    population: new FormControl('', Validators.required),
    region: new FormControl('', [Validators.required, Validators.minLength(4)]),
    flag: new FormControl('', Validators.required),
  });

  public canSave$ = this.formGroup.statusChanges.pipe(map(() => this.formGroup.valid));

  constructor(private route: ActivatedRoute, private continentService: ContinentService, private confirmation: ConfirmationService) {}

  public ngOnInit(): void {
    this.country$ = this.route.params.pipe(
      map((parameters) => parameters['country']),
      map((countryName) => this.findCountryFromContinentService(countryName)),
      switchMap((country$) => country$),

      tap((country) => this.fillInData(country))
    );
  }

  private fillInData(country: Country | undefined): void {
    console.log('Fill in data');
    this.formGroup.get('name')?.patchValue(country?.name || 'unknown');
    this.formGroup.get('region')?.patchValue(country?.region || 'unknown');
    this.formGroup.get('population')?.patchValue(country?.population || 'unknown');
    this.formGroup.get('flag')?.patchValue(country?.flag || 'unknown');
  }

  private findCountryFromContinentService(countryName: string): Observable<Country | undefined> {
    return this.continentService.countries$.pipe(map((countries) => this.findCountry(countries, countryName)));
  }

  private findCountry(countries: Country[], countryName: string): Country | undefined {
    console.log('Find country');
    const country = countries.find((c) => c.name === countryName);
    return country;
  }

  public save(): void {
    this.confirmation.confirm({
      key: 'confirm-save-country',
      message: 'Do you want to save the country?',
      accept: () => {
        this.doSave();
      },
    });
  }

  public doSave(): void {
    this.continentService.save(this.formGroup.value as Country);
  }
}
