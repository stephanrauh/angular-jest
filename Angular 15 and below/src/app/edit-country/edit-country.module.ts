import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { EditCountryComponent } from './edit-country.component';
import { Routes, RouterModule } from '@angular/router';
import { ReactiveFormsModule } from '@angular/forms';
import { ConfirmDialogModule } from 'primeng/confirmdialog';
import { ConfirmationService } from 'primeng/api';
import { InputTextModule } from 'primeng/inputtext';
import { ButtonModule } from 'primeng/button';

const routes: Routes = [{ path: ':country', component: EditCountryComponent }];

@NgModule({
  declarations: [EditCountryComponent],
  providers: [ConfirmationService],
  imports: [CommonModule, RouterModule.forChild(routes), ReactiveFormsModule, ConfirmDialogModule, InputTextModule, ButtonModule],
})
export class EditCountryModule {}
