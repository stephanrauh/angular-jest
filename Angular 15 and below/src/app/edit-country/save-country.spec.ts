import { HttpClientTestingModule } from "@angular/common/http/testing";
import { ComponentFixture, TestBed } from "@angular/core/testing";
import { NoopAnimationsModule } from "@angular/platform-browser/animations";
import { RouterTestingModule } from "@angular/router/testing";
import { ConfirmationService } from "primeng/api";
import { ConfirmDialogModule } from "primeng/confirmdialog";
import { EditCountryComponent } from "./edit-country.component";

describe('Save-Country-Question', () => {
  let fixture: ComponentFixture<EditCountryComponent>;
  let service: ConfirmationService;
  let component: EditCountryComponent;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [EditCountryComponent],
      providers: [ConfirmationService],
      imports: [ConfirmDialogModule, RouterTestingModule, HttpClientTestingModule, NoopAnimationsModule],
    }).compileComponents();

    service = TestBed.inject(ConfirmationService);

    fixture = TestBed.createComponent(EditCountryComponent);
    component = fixture.componentInstance;

    fixture.detectChanges(); // ngOnInit()
  });

  it("Should render the confirmation dialog", () => {
    service.confirm({
      key: 'confirm-save-country',
    });
    fixture.detectChanges();
    const msgSpan = document.querySelector(".p-confirm-dialog-message") as HTMLElement;
    expect(msgSpan.innerHTML).toBe("Do you want to save the country?");
    expect(document.body).toMatchSnapshot();
  })
});
