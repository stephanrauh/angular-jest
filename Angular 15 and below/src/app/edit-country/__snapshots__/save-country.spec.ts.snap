// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Save-Country-Question Should render the confirmation dialog 1`] = `
<body>
  <div
    id="root0"
  >
    <p-confirmdialog
      acceptlabel="I might!"
      class="p-element"
      header="Save country"
      key="confirm-save-country"
      message="Do you want to save the country?"
      rejectlabel="No. That's too much work."
    >
      <div
        class="p-dialog-mask p-component-overlay p-dialog-mask-scrollblocker"
      >
        <div
          class="ng-trigger ng-trigger-animation p-dialog p-confirm-dialog p-component ng-animating"
          style=""
        >
          
          <div
            class="p-dialog-header"
            style=""
          >
            <span
              class="p-dialog-title"
            >
              Save country
            </span>
            
            <div
              class="p-dialog-header-icons"
            >
              <button
                class="p-dialog-header-icon p-dialog-header-close p-link"
                type="button"
              >
                <span
                  class="pi pi-times"
                />
              </button>
              
            </div>
          </div>
          
          <div
            class="p-dialog-content"
          >
            
            <span
              class="p-confirm-dialog-message"
            >
              Do you want to save the country?
            </span>
          </div>
          
          <div
            class="p-dialog-footer"
            style=""
          >
            <button
              class="p-element p-ripple p-confirm-dialog-reject p-button p-component"
              pbutton=""
              pripple=""
              type="button"
            >
              <span
                aria-hidden="true"
                class="p-button-icon p-button-icon-left pi pi-times"
              />
              <span
                class="p-button-label"
              >
                No. That's too much work.
              </span>
            </button>
            
            <button
              class="p-element p-ripple p-confirm-dialog-accept p-button p-component"
              pbutton=""
              pripple=""
              type="button"
            >
              <span
                aria-hidden="true"
                class="p-button-icon p-button-icon-left pi pi-check"
              />
              <span
                class="p-button-label"
              >
                I might!
              </span>
            </button>
            
          </div>
          
        </div>
        
      </div>
      
    </p-confirmdialog>
    <div
      class="padding"
    >
      <form>
        <div
          class="p-grid"
        >
          <h2
            class="p-col-12"
          >
            
          </h2>
          <div
            class="p-col-2"
          >
            Name
          </div>
          <div
            class="p-col-10"
          >
            <input
              formcontrolname="name"
              name="name"
              pinputtext=""
            />
          </div>
          <small
            class="error p-col-12 p-offset-2"
          >
             The field is required. 
          </small>
          
          <div
            class="p-col-2"
          >
            Population
          </div>
          <div
            class="p-col-10"
          >
            <input
              formcontrolname="population"
              name="population"
              pinputtext=""
            />
          </div>
          <small
            class="error p-col-12 p-offset-2"
          >
             The field is required. 
          </small>
          
          <div
            class="p-col-2"
          >
            Region
          </div>
          <div
            class="p-col-10"
          >
            <input
              formcontrolname="region"
              name="region"
              pinputtext=""
            />
          </div>
          <small
            class="error p-col-12 p-offset-2"
          >
             The field is required. 
          </small>
          
          
          <div
            class="p-col-2"
          >
            Flag
          </div>
          <div
            class="p-col-10"
          >
            <input
              formcontrolname="flag"
              name="flag"
              pinputtext=""
            />
          </div>
          <small
            class="error p-col-12 p-offset-2"
          >
             The field is required. 
          </small>
          
        </div>
        <button
          class="p-element p-button p-component"
          disabled=""
          pbutton=""
        >
           Save 
        </button>
      </form>
    </div>
  </div>
</body>
`;
